ddl/ddl_api.go

```diff
@@ -3292,19 +3276,78 @@ func (d *ddl) CreateIndex(ctx sessionctx.Context, ti ast.Ident, keyType ast.Inde
 	return errors.Trace(err)
 }
 
-func buildFKInfo(fkName model.CIStr, keys []*ast.IndexColName, refer *ast.ReferenceDef, cols []*table.Column) (*model.FKInfo, error) {
+func buildFKInfo(fkName model.CIStr, keys []*ast.IndexColName, refer *ast.ReferenceDef, cols []*table.Column, tbInfo *model.TableInfo) (*model.FKInfo, error) {
 	var fkInfo model.FKInfo
 	fkInfo.Name = fkName
 	fkInfo.RefTable = refer.Table.Name
 
+	// all base columns of stored generated columns
+	baseCols := make(map[string]struct{})
+	for _, col := range cols {
+		if col.IsGenerated() && col.GeneratedStored {
+			for name := range col.Dependences {
+				baseCols[name] = struct{}{}
+			}
+		}
+	}
+
+	if len(keys) == 0 {
+		// TODO: In MySQL, this case will report a parse error.
+		return nil, infoschema.ErrCannotAddForeign
+	}
 	fkInfo.Cols = make([]model.CIStr, len(keys))
+
 	for i, key := range keys {
+		// Check add foreign key to generated columns
+		// For more detail, see https://dev.mysql.com/doc/refman/8.0/en/innodb-foreign-key-constraints.html#innodb-foreign-key-generated-columns
+		for _, col := range cols {
+			if col.Name.L != key.Column.Name.L {
+				continue
+			}
+			if col.IsGenerated() {
+				// Check foreign key on virtual generated columns
+				if !col.GeneratedStored {
+					return nil, infoschema.ErrCannotAddForeign
+				}
+
+				// Check wrong reference options of foreign key on stored generated columns
+				switch refer.OnUpdate.ReferOpt {
+				case ast.ReferOptionCascade:
+					return nil, errWrongFKOptionForGeneratedColumn.GenWithStackByArgs("ON UPDATE CASCADE")
+				case ast.ReferOptionSetNull:
+					return nil, errWrongFKOptionForGeneratedColumn.GenWithStackByArgs("ON UPDATE SET NULL")
+				case ast.ReferOptionSetDefault:
+					return nil, errWrongFKOptionForGeneratedColumn.GenWithStackByArgs("ON UPDATE SET DEFAULT")
+				}
+				switch refer.OnDelete.ReferOpt {
+				case ast.ReferOptionSetNull:
+					return nil, errWrongFKOptionForGeneratedColumn.GenWithStackByArgs("ON DELETE SET NULL")
+				case ast.ReferOptionSetDefault:
+					return nil, errWrongFKOptionForGeneratedColumn.GenWithStackByArgs("ON DELETE SET DEFAULT")
+				}
+				continue
+			}
+			// Check wrong reference options of foreign key on base columns of stored generated columns
+			if _, ok := baseCols[col.Name.L]; ok {
+				switch refer.OnUpdate.ReferOpt {
+				case ast.ReferOptionCascade, ast.ReferOptionSetNull, ast.ReferOptionSetDefault:
+					return nil, infoschema.ErrCannotAddForeign
+				}
+				switch refer.OnDelete.ReferOpt {
+				case ast.ReferOptionCascade, ast.ReferOptionSetNull, ast.ReferOptionSetDefault:
+					return nil, infoschema.ErrCannotAddForeign
+				}
+			}
+		}
 		if table.FindCol(cols, key.Column.Name.O) == nil {
 			return nil, errKeyColumnDoesNotExits.GenWithStackByArgs(key.Column.Name)
 		}
 		fkInfo.Cols[i] = key.Column.Name
 	}
 
+	if len(keys) != len(refer.IndexColNames) {
```

Comment:
Please move this logic to the beginning of the function because of the former logic does not change the value of both `keys` and `refer.IndexColNames`.